### PR TITLE
Merge `f/model-serving` to `main`

### LINKED ISSUE #1810: [Tracker]: Model Serving v2 - KServe Support
## Description

We'd like to support the ability to host KServe Model Serving servers... this differs from Model Mesh (our current structure).

## Target Branch

`f/model-serving`

## Requirements

* [Model serving phase 2 requirements](https://docs.google.com/document/d/1TXLEyzpYX6inMHOlaUW8VbMxMvQ5Y9TEzHM87G9_1Yk/edit?usp=sharing) (I'm gonna update the requirements later today)

1. Model serving runtimes
   * P0: OOTB support Caikit/TGIS
   * P0: (depending on feasibility): NVidia Triton Inference Server
   * P0: Control that some Model Serving Runtimes can only be enabled in kserve

2. Support options for KServe and/or ModelMesh (support KServe - 1 model per pod or modelmesh - multiple models per pod) 

   * P0: RHODS admins should be able to configure whether they want to use KServe (single model serving + additional functionality), ModelMesh, or both
   * P0: Users should be able to select kserve or modelmesh on project creation
      * Only one can be enabled per project
   * P0: Users should be able to deploy in the global view to both kserve and modelmesh  

## Itemized UX Issues

* https://github.com/opendatahub-io/odh-dashboard/issues/1809

### General configuration

* https://github.com/opendatahub-io/odh-dashboard/issues/1939
* https://github.com/opendatahub-io/odh-dashboard/issues/1946
* https://github.com/opendatahub-io/odh-dashboard/issues/1975

### Custom Serving Runtimes

* https://github.com/opendatahub-io/odh-dashboard/issues/1912
* https://github.com/opendatahub-io/odh-dashboard/issues/1911
* https://github.com/opendatahub-io/odh-dashboard/issues/1940
* https://github.com/opendatahub-io/odh-dashboard/issues/1973
* https://github.com/opendatahub-io/odh-dashboard/issues/1989

### Admin Settings

* https://github.com/opendatahub-io/odh-dashboard/issues/1913

### Projects

* https://github.com/opendatahub-io/odh-dashboard/issues/1941
* https://github.com/opendatahub-io/odh-dashboard/issues/1944
* https://github.com/opendatahub-io/odh-dashboard/issues/1942
* https://github.com/opendatahub-io/odh-dashboard/issues/1945
* https://github.com/opendatahub-io/odh-dashboard/issues/1943
* https://github.com/opendatahub-io/odh-dashboard/issues/1947
* https://github.com/opendatahub-io/odh-dashboard/issues/1995

### Global View

* https://github.com/opendatahub-io/odh-dashboard/issues/1948

### Enhancements

* https://github.com/opendatahub-io/odh-dashboard/issues/1579
* https://github.com/opendatahub-io/odh-dashboard/issues/1581
* https://github.com/opendatahub-io/odh-dashboard/issues/1583
* https://github.com/opendatahub-io/odh-dashboard/issues/1626
* https://github.com/opendatahub-io/odh-dashboard/issues/1353
* https://github.com/opendatahub-io/odh-dashboard/issues/1998
* https://github.com/opendatahub-io/odh-dashboard/issues/2050 

---

### Incubation work

* Unit testing coverage
* Integration testing coverage
* https://github.com/opendatahub-io/odh-dashboard/issues/1995
* https://github.com/opendatahub-io/odh-dashboard/issues/1994
* https://github.com/opendatahub-io/odh-dashboard/issues/1956
* https://github.com/opendatahub-io/odh-dashboard/issues/2053
* https://github.com/opendatahub-io/odh-dashboard/issues/2049
* https://github.com/opendatahub-io/odh-dashboard/issues/2058
* https://github.com/opendatahub-io/odh-dashboard/issues/2059
* https://github.com/opendatahub-io/odh-dashboard/issues/2048
* https://github.com/opendatahub-io/odh-dashboard/issues/2079

### Incubation Enhancements (can be done post-MVP)
* https://github.com/opendatahub-io/odh-dashboard/issues/2105
* https://github.com/opendatahub-io/odh-dashboard/issues/1580 
* https://github.com/opendatahub-io/odh-dashboard/issues/1727
* https://github.com/opendatahub-io/odh-dashboard/issues/1584
* https://github.com/opendatahub-io/odh-dashboard/issues/2047
* https://github.com/opendatahub-io/odh-dashboard/issues/2052
* https://github.com/opendatahub-io/odh-dashboard/issues/2051
* https://github.com/opendatahub-io/odh-dashboard/issues/2097
* https://github.com/opendatahub-io/odh-dashboard/issues/1980
* https://github.com/opendatahub-io/odh-dashboard/issues/2132
* https://github.com/opendatahub-io/odh-dashboard/issues/2134

## Related artifacts

* [Serving runtimes settings mockups](https://www.sketch.com/s/45df001b-2441-4124-a436-5ae1abb409c9/p/B2327543-5B01-4CF8-8BEB-AB3294F9B2C8/canvas)
* [Cluster settings mockups](https://www.sketch.com/s/45df001b-2441-4124-a436-5ae1abb409c9/p/FF910014-D6AF-4040-8343-5FBA0E9D8921/canvas)
* [Model Serving Project mockups](https://www.sketch.com/s/113593f8-5970-49d6-a352-709b07639127/p/7639AE0A-CD1D-4173-B63C-4BFB01613FD6/canvas)
* [Model Serving v2 Open Discussion](https://docs.google.com/document/d/1izyc2sjMYcIcxLpY1MlMnpXUN_Xx1i7zw7GEH5Be4Zo/edit)
* [Model Serving Controller Requirements](https://github.com/opendatahub-io/kserve/issues/92)

### PR DESCRIPTION
**Pull request by lucferbux**
## Description

Closes #1810

### CODE REVIEW

File: frontend/src/pages/modelServing/customServingRuntimes/CustomServingRuntimePlatformsLabelGroup.tsx [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  const platforms = getEnabledPlatformsFromTemplate(template);
+
+  if (platforms.length === 0) {
+    return null;
+  }
+
+  return (
+    <LabelGroup>
+      {platforms.map((platform, i) => (
+        <Label key={i}>{ServingRuntimePlatformLabels[platform]}</Label>
```
Comment (andrewballantyne) [inline frontend/src/pages/modelServing/customServingRuntimes/CustomServingRuntimePlatformsLabelGroup.tsx:28]: Small nit... it's not a huge thing usually to worry about this... but `key` has rendering value (mainly for performance). If you have rendered text like labels, you probably could just use that as the key. If that's not unique, the user will likely be confused too.

Unless we are rendering a ton of items, it typically won't matter one way or another. But I like to think it gives the developer more understanding of React to treat `key`s with the intent they have.

File: frontend/src/pages/modelServing/screens/global/InferenceServiceListView.tsx [hunk 2]
```diff
@@ -5,8 +5,9 @@ import SearchField, { SearchType } from '~/pages/projects/components/SearchField
 import { ModelServingContext } from '~/pages/modelServing/ModelServingContext';
 import { ProjectsContext } from '~/concepts/projects/ProjectsContext';
 import { getInferenceServiceDisplayName, getInferenceServiceProjectDisplayName } from './utils';
-import ServeModelButton from './ServeModelButton';
+//import ServeModelButton from './ServeModelButton';
```
Comment (andrewballantyne) [inline frontend/src/pages/modelServing/screens/global/InferenceServiceListView.tsx:8]: Commented code

File: frontend/src/pages/modelServing/screens/global/ModelServingProjectSelection.tsx [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    <SplitItem>
+      <Bullseye>Project</Bullseye>
+    </SplitItem>
+    <SplitItem>
+      {/* Maybe we want to filter the projects with no deployed models that's why I added the filterLable prop */}
+      <ProjectSelectorNavigator
+        getRedirectPath={getRedirectPath}
+        invalidDropdownPlaceholder={'All projects'}
+        selectAllProjects
+        filterLabel={KnownLabels.DASHBOARD_RESOURCE}
```
Comment (andrewballantyne) [inline frontend/src/pages/modelServing/screens/global/ModelServingProjectSelection.tsx:23]: Is this necessary? Why would we ever want to filter this list down?

File: frontend/src/pages/projects/screens/detail/DetailsSection.tsx [hunk 4]
```diff
@@ -17,9 +17,10 @@ type DetailsSectionProps = {
   title: string;
   isLoading: boolean;
   loadError?: Error;
-  isEmpty: boolean;
-  emptyState: React.ReactNode;
+  isEmpty?: boolean;
+  emptyState?: React.ReactNode;
```
Comment (andrewballantyne) [inline frontend/src/pages/projects/screens/detail/DetailsSection.tsx:21]: Was this change necessary? 🤔 I think it has value to always ask this question as the intent is for you to handle "empty" states...

File: frontend/src/services/getClusterStatusService.ts
Comment (andrewballantyne) [inline frontend/src/services/getClusterStatusService.ts:1]: Dead code, please remove.

File: frontend/src/services/templateService.ts [hunk 5]
```diff
@@ -78,6 +82,19 @@ export const updateServingRuntimeTemplateBackend = (
...
+                path: '/metadata/annotations/opendatahub.io~1modelServingSupport',
+                value: JSON.stringify(platforms),
+              }
+            : {
+                op: 'add',
+                path: '/metadata/annotations',
+                value: {
+                  'opendatahub.io/modelServingSupport': JSON.stringify(platforms),
+                },
+              },
```
Comment (andrewballantyne) [inline frontend/src/services/templateService.ts:97]: Dupe code. `frontend/src/api/k8s/templates.ts` -- we should resolve this... it'll be a maintenance annoyance.

**Review by andrewballantyne — COMMENTED** [review]
Questions I would like answered asap:
* https://github.com/opendatahub-io/odh-dashboard/pull/2123#discussion_r1391108510
* https://github.com/opendatahub-io/odh-dashboard/pull/2123#discussion_r1391111322
* https://github.com/opendatahub-io/odh-dashboard/pull/2123#discussion_r1391142458

The rest is followup / cleanup. Log tickets as needed to work through things.

File: frontend/src/api/apiMergeUtils.ts [hunk 6]
```diff
@@ -1,6 +1,13 @@
-import { K8sResourceBaseOptions, QueryParams } from '@openshift/dynamic-plugin-sdk-utils';
+import {
+  K8sResourceBaseOptions,
+  K8sResourceDeleteOptions,
+  QueryParams,
+} from '@openshift/dynamic-plugin-sdk-utils';
 import { K8sAPIOptions } from '~/k8sTypes';
 
+const dryRunPayload = (dryRun?: boolean): Pick<K8sResourceDeleteOptions, 'payload'> =>
+  dryRun ? { payload: { dryRun: ['All'] } } : {};
```
[the remark this answers is not in the document]
Comment (christianvogt) [reply frontend/src/api/apiMergeUtils.ts:9]: The `payload` property is specific to dry run for `DELETE` where the existing functionality doesn't apply to `DELETE`.

File: frontend/src/components/ResourceActionsColumn.tsx
[the remark this answers is not in the document]
Comment (christianvogt) [reply frontend/src/components/ResourceActionsColumn.tsx:1]: Yes it's table specific and probably should be put into the `table` dir.

File: manifests/modelserving/kustomization.yaml [hunk 7]
```diff
@@ -4,9 +4,15 @@ commonLabels:
...
     newName: quay.io/opendatahub/openvino_model_server
-    digest: sha256:20dbfbaf53d1afbd47c612d953984238cb0e207972ed544a5ea662c2404f276d
\ No newline at end of file
+    digest: sha256:2cbe8a48ab0bc6fe7fb76919bf33253e83a6218a9c4b486b744c3dcf30679616
+  - name: caikit-tgis-serving
+    newName: quay.io/opendatahub/caikit-tgis-serving
+    digest: sha256:9d6e8bc10f3b61774f9d321790f8593f6ded5674b64a366248fcde6a07d66833
+  - name: text-generation-inference
+    newName: quay.io/opendatahub/text-generation-inference
+    digest: sha256:54959e3dc0e3bb139e9db5b912143a7dedcef725c55db58cfef0e71dd057e4a5
```
Comment (andrewballantyne) [inline manifests/modelserving/kustomization.yaml:18]:
```suggestion
    digest: sha256:efcc60c671a5da62c0251bc543f6de73f6578aac40a9a745eae3f60b0208f9e9
```
cc @Xaenalt

Comment (andrewballantyne) [reply manifests/modelserving/kustomization.yaml:18]: This will be a follow up -- I don't believe this will be done to this PR.

File: frontend/src/const.ts [hunk 8]
```diff
@@ -1,6 +1,6 @@
 import { KnownLabels } from '~/k8sTypes';
 
 export const LABEL_SELECTOR_DASHBOARD_RESOURCE = `${KnownLabels.DASHBOARD_RESOURCE}=true`;
-export const LABEL_SELECTOR_MODEL_SERVING_PROJECT = `${KnownLabels.MODEL_SERVING_PROJECT}=true`;
+export const LABEL_SELECTOR_MODEL_SERVING_PROJECT = KnownLabels.MODEL_SERVING_PROJECT;
```
[the remark this answers is not in the document]
Comment (christianvogt) [reply frontend/src/const.ts:4]: Technically it is still a selector but it accepts _any_ value so long as the label is present. Probably could have used a comment and / or renamed the const to be more precise.

File: frontend/src/api/k8s/inferenceServices.ts [hunk 9]
```diff
@@ -35,7 +36,13 @@ const assembleInferenceService = (
...
       annotations: {
         'openshift.io/display-name': data.name.trim(),
-        'serving.kserve.io/deploymentMode': 'ModelMesh',
+        ...(isModelMesh
+          ? { 'serving.kserve.io/deploymentMode': 'ModelMesh' }
+          : {
+              'serving.knative.openshift.io/enablePassthrough': 'true',
+              'sidecar.istio.io/inject': 'true',
+              'sidecar.istio.io/rewriteAppHTTPProbers': 'true',
+            }),
```
[the remark this answers is not in the document]
Comment (bartoszmajsak) [reply frontend/src/api/k8s/inferenceServices.ts:45]: > Service Mesh uses this value and they only do it in Projects that they have annotated

I think we are mixing two "service mesh use cases" here (and another mesh being model mesh (sigh!) :D). 

The annotation on the projects, which you have in mind (`opendatahub.io/service-mesh`), is aimed to support/enable new Auth(z) flow using Service Mesh and Authorino. I think this is entirely unrelated (at least at this point in time) to what Model Mesh needs as far as configuring resources is concerned. If I understand this piece of code correctly, here you set up `InferenceService` with certain Istio features required by it. I do not know the details of why those are needed (except of `'sidecar.istio.io/inject': 'true'`), so perhaps a small comment giving reasons would clear the confusion. HTH

[the remark this answers is not in the document]
Comment (bartoszmajsak) [reply frontend/src/api/k8s/inferenceServices.ts:45]: One more thing worth adding - there's no controller in the ODH deployment yet which can act on `opendatahub.io/service-mesh` annotation.

[the remark this answers is not in the document]
Comment (bartoszmajsak) [reply frontend/src/api/k8s/inferenceServices.ts:45]: One more note to clarify: as this change is targeting `main` branch and we do not have a logic yet to add `opendatahub.io/service-mesh` (behind `disableServiceMesh` flag) those changes are not handled by any controller in the cluster.

[the remark this answers is not in the document]
Comment (israel-hdez) [reply frontend/src/api/k8s/inferenceServices.ts:45]: For the time being, the `odh-model-controller` will take care of adding a namespace/project to the service mesh when an InferenceService for KServe is created. So, these annotations are, indeed, required for InferenceServices to work correctly with KServe.

File: manifests/modelserving/kustomization.yaml (hunk 7 above)
Comment (andrewballantyne) [reply manifests/modelserving/kustomization.yaml:18]: Fixed by https://github.com/opendatahub-io/odh-dashboard/pull/2128

File: frontend/src/pages/projects/screens/detail/DetailsSection.tsx (hunk 4 above)
Comment (DaoDaoNoCode) [reply frontend/src/pages/projects/screens/detail/DetailsSection.tsx:21]: I changed the interface at first but then realized I needed them but forgot to change them back. This should be addressed in the follow-up fixes.

**Review by andrewballantyne — APPROVED** [review]

File: frontend/src/pages/modelServing/screens/global/ModelServingProjectSelection.tsx (hunk 3 above)
Comment (DaoDaoNoCode) [reply frontend/src/pages/modelServing/screens/global/ModelServingProjectSelection.tsx:23]: Will be fixed in https://github.com/opendatahub-io/odh-dashboard/pull/2154

File: frontend/src/pages/projects/screens/detail/DetailsSection.tsx (hunk 4 above)
Comment (DaoDaoNoCode) [reply frontend/src/pages/projects/screens/detail/DetailsSection.tsx:21]: Will be fixed in https://github.com/opendatahub-io/odh-dashboard/pull/2154

File: frontend/src/pages/modelServing/screens/global/InferenceServiceListView.tsx (hunk 2 above)
Comment (DaoDaoNoCode) [reply frontend/src/pages/modelServing/screens/global/InferenceServiceListView.tsx:8]: Will be fixed in https://github.com/opendatahub-io/odh-dashboard/pull/2154

File: frontend/src/k8sTypes.ts [hunk 10]
```diff
@@ -291,12 +294,14 @@ export type PodKind = K8sResourceCommon & {
   };
 };
 
+/** Assumed Dashboard Project -- if we need more beyond that we should break this type up */
```
[the remark this answers is not in the document]
Comment (DaoDaoNoCode) [reply frontend/src/k8sTypes.ts:297]: Will be fixed in https://github.com/opendatahub-io/odh-dashboard/pull/2154

File: frontend/src/concepts/projects/ProjectSelector.tsx [hunk 11]
```diff
@@ -39,17 +47,32 @@ const ProjectSelector: React.FC<ProjectSelectorProps> = ({
...
-        </DropdownItem>
-      ))}
+      dropdownItems={[
+        ...(selectAllProjects
+          ? [
+              <DropdownItem
+                key={'all-projects'}
+                onClick={() => {
+                  setDropdownOpen(false);
+                  onSelection({ metadata: { name: '' } } as ProjectKind);
```
[the remark this answers is not in the document]
Comment (DaoDaoNoCode) [reply frontend/src/concepts/projects/ProjectSelector.tsx:57]: Will be fixed in https://github.com/opendatahub-io/odh-dashboard/pull/2154

File: frontend/src/__tests__/integration/pages/modelServing/ModelServingGlobal.stories.tsx [hunk 12]
```diff
@@ -3,8 +3,9 @@ import React from 'react';
 import { StoryFn, Meta, StoryObj } from '@storybook/react';
 import { rest } from 'msw';
 import { within, userEvent } from '@storybook/testing-library';
-import { expect } from '@storybook/jest';
+// import { expect } from '@storybook/jest';
```
[the remark this answers is not in the document]
Comment (DaoDaoNoCode) [reply frontend/src/__tests__/integration/pages/modelServing/ModelServingGlobal.stories.tsx:6]: Will be fixed in https://github.com/opendatahub-io/odh-dashboard/pull/2154
